src/main/java/org/opensearch/ad/transport/ForwardADTaskTransportAction.java

```diff
@@ -26,45 +26,145 @@
 
     @Inject
-    public ForwardADTaskTransportAction(ActionFilters actionFilters, TransportService transportService, ADTaskManager adTaskManager) {
+    public ForwardADTaskTransportAction(
+        ActionFilters actionFilters,
+        TransportService transportService,
+        ADTaskManager adTaskManager,
+        ADTaskCacheManager adTaskCacheManager
+    ) {
         super(ForwardADTaskAction.NAME, transportService, actionFilters, ForwardADTaskRequest::new);
         this.adTaskManager = adTaskManager;
         this.transportService = transportService;
+        this.adTaskCacheManager = adTaskCacheManager;
     }
 
     @Override
     protected void doExecute(Task task, ForwardADTaskRequest request, ActionListener<AnomalyDetectorJobResponse> listener) {
         ADTaskAction adTaskAction = request.getAdTaskAction();
         AnomalyDetector detector = request.getDetector();
         DetectionDateRange detectionDateRange = request.getDetectionDateRange();
+        String detectorId = detector.getDetectorId();
+        ADTask adTask = request.getAdTask();
 
         switch (adTaskAction) {
             case START:
+                // Start historical analysis for detector
                 adTaskManager.startHistoricalAnalysisTask(detector, detectionDateRange, request.getUser(), transportService, listener);
                 break;
-            case STOP:
+            case FINISHED:
+                // Historical analysis finished, so we need to remove detector cache.
                 adTaskManager.removeDetectorFromCache(request.getDetector().getDetectorId());
                 listener.onResponse(new AnomalyDetectorJobResponse(detector.getDetectorId(), 0, 0, 0, RestStatus.OK));
                 break;
+            case NEXT_ENTITY:
+                // Run next entity for HC detector historical analysis.
+                logger.debug("Received task for NEXT_ENTITY action: {}", adTask.getTaskId());
+                if (detector.isMultientityDetector()) { // AD task could be HC detector level task or entity task
+                    adTaskCacheManager.removeRunningEntity(detectorId, adTask.getEntity().get(0).getValue());
+                    if (!adTaskCacheManager.hasEntity(detectorId)) {
+                        logger.info("Historical HC detector done, will remove from cache, detector id:{}", detectorId);
+                        listener.onResponse(new AnomalyDetectorJobResponse(detectorId, 0, 0, 0, RestStatus.OK));
+                        // TODO: reset task state when get task
+                        ADTaskState state = !adTask.isEntityTask() && adTask.getError() != null ? ADTaskState.FAILED : ADTaskState.FINISHED;
+                        adTaskManager.setHCDetectorTaskDone(adTask, state, listener);
+                    } else {
+                        logger.debug("Run next entity for detector " + detectorId);
+                        adTaskManager.runBatchResultActionForEntity(adTask, listener);
```

Comment:
The batch means we query feature data by batch and batch result action introduced since last release which only support single entity detector. 
Batch result action is named according to the original AD result action which only support get feature data in one detection interval.

Reply:
Will change to `runNextEntityForHCADHistorical` if that make it easier for you to understand.

Reply:
yes, it will.  thanks.